### PR TITLE
ESQL: extend Validatable support

### PR DESCRIPTION
**Pull request by ivancea**
- Moved `TopList` agg foldable validations to `Validatable`. Same for other functions doing the same.
- Add `Validatable` support in function tests
- Add `Validatable` to functions documentation
- Fixed `Bucket` `Validatable` implementation to not fail test cases with `null` parameters

## Do not merge
Waiting for https://github.com/elastic/elasticsearch/issues/100634

### CODE REVIEW

File: x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/TypeResolutions.java [hunk 1]
```diff
@@ -117,58 +117,6 @@ public static TypeResolution isIPAndExact(Expression e, String operationName, Pa
         return isExact(e, operationName, paramOrd);
     }
 
-    public static TypeResolution isFoldable(Expression e, String operationName, ParamOrdinal paramOrd) {
```
Comment (ivancea) [inline x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/TypeResolutions.java:120]: As this shouldn't be used on type resolution, I don't think we should have it here. QL has it here, but now with esql-core we can be more strict

**Review by astefan — COMMENTED** [review]
I don't think the changes in this PR should be done without also considering the rest of the hypothetical work in this area https://github.com/elastic/elasticsearch/issues/100634. Instead I would add this task of extending the aggregations testing infrastructure to the issue above and to be done after the other items are finished.

**Comment by ivancea** [comment]
> I don't think the changes in this PR should be done without also considering the rest of the hypothetical work in this area #100634. Instead I would add this task of extending the aggregations testing infrastructure to the issue above and to be done after the other items are finished.

@astefan Indeed, looks like something that complements that issue. However, I would still go with this, for some reasons:
- validate() was an existing thing, currently used in other 2 functions. So I consider this adding missing testing support to an existing feature
  - I've seen [this other PR](https://github.com/elastic/elasticsearch/pull/110021) that could benefit from testing validations. So the sooner we add this, the sooner we will be able to have such tests "standardized"
- This PR is removing the possibility to check foldability in `resolveType()`. So we won't have to migrate more things in the future
- I don't see a clear order requirement with that issue. It would make sense to have made this at the end, but also I don't see a reason to not just merge this now and have it ready

**Comment by astefan** [comment]
> > I don't think the changes in this PR should be done without also considering the rest of the hypothetical work in this area #100634. Instead I would add this task of extending the aggregations testing infrastructure to the issue above and to be done after the other items are finished.
> 
> @astefan Indeed, looks like something that complements that issue. However, I would still go with this, for some reasons:
> 
> * validate() was an existing thing, currently used in other 2 functions. So I consider this adding missing testing support to an existing feature
>   
>   * I've seen [this other PR](https://github.com/elastic/elasticsearch/pull/110021) that could benefit from testing validations. So the sooner we add this, the sooner we will be able to have such tests "standardized"
> * This PR is removing the possibility to check foldability in `resolveType()`. So we won't have to migrate more things in the future
> * I don't see a clear order requirement with that issue. It would make sense to have made this at the end, but also I don't see a reason to not just merge this now and have it ready

I don't feel comfortable adding code (this PR) for something that's not yet fixed in its entirety (the whole story of constants support in aggregations, with adjusting the optimization rule(s) if needed, implementing this support for all of them). This PR just assumes the future will look in a certain way and it's getting ahead of it.
It's a more unitary work that doesn't let room for unnecessary/unused code if the future turns out to be different.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java [hunk 2]
```diff
@@ -83,14 +83,6 @@ public void testAggsExpressionsInStatsAggs() {
                 + " found value [first_name] type [keyword]",
             error("from test | stats count(avg(first_name)) by first_name")
         );
-        assertEquals(
```
[the remark this answers is not in the document]
Comment (alex-spies) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java:86]: These test cases are not part of a fix; the second argument of `percentile` and `count_distinct` must be a constant, and we should test the corresponding validation. I'm not opposed to moving them into agg-function specific test classes - as long as there's at least one test that ensures that the agg-function specific validations are called after the optimizer has run.

**Pull request closed without merging**